### DISCUSSION TITLE
want to create a demo to print request response payloads

### DISCUSSION THREAD

**Discussion by Rajdeep-Bhattacharya** [open-telemetry/opentelemetry-java-instrumentation#9491] [Q&A]
im trying to create a demo where i can run the java agent with one of my services, which is built on spring boot, and collect traces to zipkin i also want to add the request response payloads for the apis of my service without making client changes
what is the best way to go about doing this? 
i thought i could look at the[ java-http-client](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/java-http-client) instrumentation implementation and create my own instrumentation override which can add these fields but i dont see the prints or the attributes added to the spans 

is this the correct way to proceed or should i be looking somewhere else?

let me know if i need to provide more details

```java 
package io.opentelemetry.javaagent.instrumentation.myhttpclient;

import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationHelper;
import net.bytebuddy.asm.Advice;
import com.example.HttpRequest;
import java.util.Arrays;

public class MyHttpClientAdvice {
  @Advice.OnMethodEnter
  public static void onEnter(@Advice.Argument(0) HttpRequest request) {
    //System.out.println(Arrays.toString(request.bodyPublisher().stream().toArray()));
///add details to span here
  }

  @Advice.OnMethodExit(onThrowable = Throwable.class)
  public static void onExit(@Advice.Return Object response, @Advice.Thrown Throwable throwable) {

  }
}
```

```java
package io.opentelemetry.javaagent.instrumentation.myhttpclient;

import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationHelper;
import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.matcher.ElementMatcher;

import static net.bytebuddy.matcher.ElementMatchers.not;
import static net.bytebuddy.matcher.ElementMatchers.named;

public class MyHttpClientTypeInstrumentation implements TypeInstrumentation {
  @Override
  public ElementMatcher<TypeDescription> typeMatcher() {
    return nameStartsWith("java.net.")
        .or(nameStartsWith("jdk.internal."))
        .and(not(named("jdk.internal.net.http.HttpClientFacade")))
        .and(extendsClass(named("java.net.http.HttpClient")));
  }

  @Override
  public void transform(TypeTransformer transformer) {
    transformer.applyAdviceToMethod(
        isMethod().and(named("send")).and(takesArguments(1)).and(takesArgument(0, named("java.net.http.HttpClient"))),
        MyHttpClientAdvice.class.getName());
  }
}
```

```java
package io.opentelemetry.javaagent.instrumentation.myhttpclient;

import com.google.auto.service.AutoService;
import io.opentelemetry.javaagent.tooling.InstrumentationModule;
import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
import java.util.Collections;
import java.util.List;

@AutoService(InstrumentationModule.class)
public class MyHttpClientInstrumentationModule extends InstrumentationModule {
  public MyHttpClientInstrumentationModule() {
    super("my-http-client", "my-http-client-1.0");
  }

  @Override
  public List<TypeInstrumentation> typeInstrumentations() {
    return Collections.singletonList(new MyHttpClientTypeInstrumentation());
  }
}
```

**Comment by Rajdeep-Bhattacharya** [comment]
other things ive tried 
1. setting a debugger and running my service with the otel repo i can see the breakpoints at the attribute getter code like [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/331aa04e351b46c57181fd28d0a3e6aabd15d880/instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatHttpAttributesGetter.java#L54C3-L54C3) but i dont understand where the code goes from there and where these attributes are being populated in the span
2. i tried adding code in [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/331aa04e351b46c57181fd28d0a3e6aabd15d880/instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatHelper.java#L65) in the hopes i can print the inputStream in the coyote REQUEST object but any code i write here needs to have a unit test associated and needs to follow checkStyles constraints which if i ignore in the build command excludes the tomcat instrumentation all together

**Comment by trask** [comment]
hi @Rajdeep-Bhattacharya! check out the related #8778

**Reply by Rajdeep-Bhattacharya** [reply]
thank you @trask
